component/azstorage/block_blob.go

```diff
@@ -766,16 +772,98 @@ func (bb *BlockBlob) createNewBlocks(blockList *common.BlockOffsetList, offset,
 }
 
+func (bb *BlockBlob) removeBlocks(blockList *common.BlockOffsetList, size int64, name string) *common.BlockOffsetList {
+	_, index := blockList.BinarySearch(size)
+	// if the start index is equal to new size - block should be removed - move one index back
+	if blockList.BlockList[index].StartIndex == size {
+		index = index - 1
+	}
+	// if the file we're shrinking is in the middle of a block then shrink that block
+	if blockList.BlockList[index].EndIndex > size {
+		blk := blockList.BlockList[index]
+		blk.EndIndex = size
+		blk.Data = make([]byte, blk.EndIndex-blk.StartIndex)
+		blk.Flags.Set(common.DirtyBlock)
+		bb.ReadInBuffer(name, blk.StartIndex, blk.EndIndex-blk.StartIndex, blk.Data)
+	}
+
+	blockList.BlockList = blockList.BlockList[:index+1]
+	return blockList
+}
+
+func (bb *BlockBlob) TruncateFile(name string, size int64) error {
+	// log.Trace("BlockBlob::TruncateFile : name=%s, size=%d", name, size)
+	attr, err := bb.GetAttr(name)
+	if err != nil {
+		log.Err("BlockBlob::TruncateFile : Failed to get attributes of file %s (%s)", name, err.Error())
+		if err == syscall.ENOENT {
+			return err
+		}
+	}
+	if size == 0 || attr.Size == 0 {
+		err := bb.WriteFromBuffer(name, nil, make([]byte, size))
+		if err != nil {
+			log.Err("BlockBlob::TruncateFile : Failed to set the %s to 0 bytes (%s)", name, err.Error())
+		}
+		return err
+	}
+	bol, err := bb.GetFileBlockOffsets(name)
+	if err != nil {
+		log.Err("BlockBlob::TruncateFile : Failed to get block list of file %s (%s)", name, err.Error())
+		return err
+	}
+	if !bol.SmallFile() {
+		if size > attr.Size {
+			bb.createNewBlocks(bol, bol.BlockList[len(bol.BlockList)-1].EndIndex, size-attr.Size)
+		} else if size < attr.Size {
+			bol = bb.removeBlocks(bol, size, name)
+		}
+		err = bb.StageAndCommit(name, bol)
+		if err != nil {
+			log.Err("BlockBlob::TruncateFile : Failed to truncate file %s", name, err.Error())
+			return err
+		}
+	} else {
+		data, err := bb.ReadBuffer(name, 0, 0)
+		if err != nil {
+			log.Err("BlockBlob::TruncateFile : Failed to read small file %s", name, err.Error())
+			return err
+		}
+		if size > attr.Size {
+			blk := &common.Block{
+				StartIndex: 0,
+				EndIndex:   attr.Size,
+				Data:       data,
+				Id:         base64.StdEncoding.EncodeToString(common.NewUUID().Bytes()),
+			}
+			blk.Flags.Set(common.DirtyBlock)
+			bol.Flags.Clear(common.SmallFile)
+
+			bol.BlockList = append(bol.BlockList, blk)
+			bol.BlockIdLength = common.GetIdLength(blk.Id)
+			bb.createNewBlocks(bol, bol.BlockList[len(bol.BlockList)-1].EndIndex, size-attr.Size)
+		} else if size < attr.Size {
+			data = data[0:size]
+			return bb.WriteFromBuffer(name, nil, data)
+		}
+		err = bb.StageAndCommit(name, bol)
+		if err != nil {
+			log.Err("BlockBlob::TruncateFile : Failed to truncate file %s", name, err.Error())
+			return err
+		}
+	}
+	return nil
+}
+
 // Write : write data at given offset to a blob
 func (bb *BlockBlob) Write(options internal.WriteFileOptions) error {
```

Comment:
We shouldnt enter a situation where a block that is being written is a TruncatedBlock right? Since we immediately StageAndCommit on Truncate?

Reply:
We shouldn't, since `write` is used when we're streaming directly, and the truncate operation is also essentially being streamed.